### ISSUE TITLE
HANA System Replication not established due to missing backup when deploying with Netweaver

### ISSUE THREAD

**Issue by diegoakechi** [SUSE/ha-sap-terraform-deployments#739]
**Used cloud platform**
Azure

**Used SLES4SAP version**
SLES 4 SAP 15-SP2

**Used client machine OS**
openSUSE TW

**Expected behaviour vs observed behaviour**
It was expected a fully functional cluster, with System Replication established.

The current result is a configured HANA cluster, with a System Replication configured but not connected:

```
prdadm@vmhana01:/usr/sap/PRD/HDB00> HDBSettings.sh systemReplicationStatus.py
| Database | Host     | Port  | Service Name | Volume ID | Site ID | Site Name | Secondary | Secondary | Secondary | Secondary | Secondary     | Replication | Replication | Replication                                                                   | 
|          |          |       |              |           |         |           | Host      | Port      | Site ID   | Site Name | Active Status | Mode        | Status      | Status Details                                                                | 
| -------- | -------- | ----- | ------------ | --------- | ------- | --------- | --------- | --------- | --------- | --------- | ------------- | ----------- | ----------- | ----------------------------------------------------------------------------- | 
| SYSTEMDB | vmhana01 | 30001 | nameserver   |         1 |       1 | Site1     | vmhana02  |     30001 |         2 | Site2     | YES           | SYNC        | ACTIVE      |                                                                               | 
| PRD      | vmhana01 | 30007 | xsengine     |         2 |       1 | Site1     | vmhana02  |     30007 |         2 | Site2     | YES           | SYNC        | ERROR       | Connection refused: Primary needs initial data backup for system replication! | 
| PRD      | vmhana01 | 30003 | indexserver  |         3 |       1 | Site1     | vmhana02  |     30003 |         2 | Site2     | YES           | SYNC        | ERROR       | Connection refused: Primary needs initial data backup for system replication! |

status system replication site "2": ERROR
overall system replication status: ERROR

Local System Replication State
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

mode: PRIMARY
site id: 1
site name: Site1
```

Running the full backup again solves the problem:

```
hdbsql SYSTEMDB=> BACKUP DATA FOR FULL SYSTEM USING FILE('BACKUP_MANUAL.BKP');
0 rows affected (overall time 118.332238 sec; server time 118.270791 sec)
```
After the backup, the System Replication is automatically established:

```
prdadm@vmhana01:/hana/backup> HDBSettings.sh systemReplicationStatus.py
| Database | Host     | Port  | Service Name | Volume ID | Site ID | Site Name | Secondary | Secondary | Secondary | Secondary | Secondary     | Replication | Replication | Replication    | 
|          |          |       |              |           |         |           | Host      | Port      | Site ID   | Site Name | Active Status | Mode        | Status      | Status Details | 
| -------- | -------- | ----- | ------------ | --------- | ------- | --------- | --------- | --------- | --------- | --------- | ------------- | ----------- | ----------- | -------------- | 
| SYSTEMDB | vmhana01 | 30001 | nameserver   |         1 |       1 | Site1     | vmhana02  |     30001 |         2 | Site2     | YES           | SYNC        | ACTIVE      |                | 
| PRD      | vmhana01 | 30007 | xsengine     |         2 |       1 | Site1     | vmhana02  |     30007 |         2 | Site2     | YES           | SYNC        | ACTIVE      |                | 
| PRD      | vmhana01 | 30003 | indexserver  |         3 |       1 | Site1     | vmhana02  |     30003 |         2 | Site2     | YES           | SYNC        | ACTIVE      |                |

status system replication site "2": ACTIVE
overall system replication status: ACTIVE

Local System Replication State
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

mode: PRIMARY
site id: 1
site name: Site1
```

I suspect it can be a race condition between the System Replication enablement (that expects a fresh backup in place) vs Netweaver installation running in parallel that introduces changes in the middle. 

**How to reproduce**
Using the default recommendations with full installation (HANA + Netweaver + Monitoring). 

**Logs**
Please, refer to the Trento Demo environment.

**Comment by arbulu89** [comment]
Hi @diegoakechi ,
Thanks for the report. About your analysis I have my doubts. The system replication is fully functioning for the moment when we start the NW installation. I'm starting to think that we miss some of these steps in our deployment, that are not coming in the official installation guide (which we follow). Maybe, we need to run a backup after the NW installation, as this brings a lot of changes and HANA requires a new backup.

PD: I guess this is happening in Netweaver 7.5, right?

**Comment by diegoakechi** [comment]
@arbulu89  I couldn't find any similar reference. For sure running a backup after the deployment solves the problem, but not sure if we would be treating the symptom here. 
Another possibility would be to disable (or only enable) SR after the NW installation. So we make the backup once after all the DB operations are finished. This would possibly speed up the DB import too. 

We should check if @pirat013, @gereonvey or @fmherschel have some insights about this issue.

**Comment by fmherschel** [comment]
@diegoakechi and @arbulu89 The  full backup of SAP HANA us needed to setup a proper system replication. It is also mentioned in our best practice guide and not a workaround.

**Comment by fmherschel** [comment]
Sorry full backup means a backup of ALL databases (SYSTEMDB and all tenant databases). If you add tenant databases you need to backup the new tenants.

**Comment by fmherschel** [comment]
In my point of we could:
a) disable SR before the NW installation; after NW installation do a full backup and setup SR again
=> This maybe also speeds-up the SR and less logs needs to be transferred and held, if the secondary would be offline
during that step.
b) Doing a backup of added tenants after the installation - need to check if that needs additional SR commands to
tell the SR that all tenants are now backed-up

**Comment by gereonvey** [comment]
By installing the NetWeaver system you added a new tenant database to the HANA system, which previously only held the SYSTEM database. In multiple-container systems, the system database and all tenant databases must be backed up. This is necessary to start creating log backups. Activated log backup is a prerequisite to get a common sync point for log shipping between the primary and secondary system.

**Comment by fmherschel** [comment]
In your installation FLOW I guess that SAP HANA gets installed with SYSTEMDB and in tenant mode but without PRD. Then the NW installation adds a complete database (PRD) which is the game-changer here. I am also a bit wondering why SAP has this role to a must-have-a-backup. An SAP colleague always says that the backup is used to transfer the initial SR data, but I am not sure. if that is still really true. However SAP defines that they need a backup of all component databases.

**Comment by diegoakechi** [comment]
Thanks a lot @fmherschel and @gereonvey .

@arbulu89 seems to be a known operational situation, so now it is up to you to figure out the best way to solve that.

**Comment by arbulu89** [comment]
Thanks @fmherschel @gereonvey , It would be good if we had this information in our installation guides (for example: https://documentation.suse.com/sbp/all/html/SAP_S4HA10_SetupGuide-SLE12/index.html).

About adding a new tenant, I don't think the Netweaver DB installation adds a new tenant, it imports the DB schema to an already existing tenant, that we actually point out during the installation parameters setting. I guess that as the change is massive, it requires there a new backup.

**Comment by fmherschel** [comment]
@arbulu89 At least I have seen SAP HANA installing in a tenant-mode and adding a tenant later was exactly matching your symptom. Why should adding a database schema a difference to a massive production? It all ends-up in log shipping. 
But adding tenants does open NEW SR channels. That a complete different thing.

**Comment by fmherschel** [comment]
@arbulu89 The guide you referenced is BY INTENTION focusing on the ASCS/ERS cluster and SR for the SAP HANA is not
discussed there as that is a complete other cluster. I would say we could add it to the performance optimized SR scenario and
add an info to the backup section.

**Comment by fmherschel** [comment]
@arbulu89 In your installation flow - is the SR active during the installation of SAP NW?
a) if yes -> I would expect that massive changes would be covered. But as I expect the tenant to be new that would not work.
    This is also why the message would ask for an initial backup
b) if no -> if the SR would be setup before but you de-register the secondary during the NW installation then you are right, then
    the massive changes could not be covered and the logs are not kept for a later log shipping. In that case it could be that you
    need a second 'initial' backup
But I still think that the NW install adds the tenant as I have seen before. Of course that depends also on how you did install SAP HANA exactly.
Maybe we could "debug" that by doing a tenant-based initial backup after the SAP HANA installation (not  a full system backup). Then we could "see", if we already have a PRD backup at the point the flow breaks with the SR problem.

**Comment by yeoldegrove** [comment]
> In your installation flow - is the SR active during the installation of SAP NW?
Yes

In our setup here, the PRD tenant is created by the HANA installation itself.
After that, a full backup (of all DBs) is done and the SR is established.

To make sure a DB import is not run before the backup+SR is done, we should really only start to deploy Netweaver when we know all HANA salt states are done.
This is not the case at the moment and salt states on HANA and Netweaver might run in parallel.
Netweaver/Import states wait for the DB to be reachable before they start... Which is also the case if SR is not yet established.

possible fix https://github.com/SUSE/ha-sap-terraform-deployments/commit/08c4c94f67fea3c690c884a3191c71f9d906e435#diff-afcb3ab4d99298cacacab676e53e1ef05f8159f8e7087a0a3f2cbe9ca32c7b7fR216

Another way to fix it would be to not query for the DB port but for the SR state.

**Comment by arbulu89** [comment]
> > In your installation flow - is the SR active during the installation of SAP NW?
> > Yes
> 
> In our setup here, the PRD tenant is created by the HANA installation itself.
> After that, a full backup (of all DBs) is done and the SR is established.
> 
> To make sure a DB import is not run before the backup+SR is done, we should really only start to deploy Netweaver when we know all HANA salt states are done.
> This is not the case at the moment and salt states on HANA and Netweaver might run in parallel.
> Netweaver/Import states wait for the DB to be reachable before they start... Which is also the case if SR is not yet established.
> 
> possible fix [08c4c94#diff-afcb3ab4d99298cacacab676e53e1ef05f8159f8e7087a0a3f2cbe9ca32c7b7fR216](https://github.com/SUSE/ha-sap-terraform-deployments/commit/08c4c94f67fea3c690c884a3191c71f9d906e435#diff-afcb3ab4d99298cacacab676e53e1ef05f8159f8e7087a0a3f2cbe9ca32c7b7fR216)
> 
> Another way to fix it would be to not query for the DB port but for the SR state.

@yeoldegrove This is not entirely sure.
Keep in mind that most of our deployments are in HA setup. We check if the HANA database is available using the cluster virtual ip address, so this means it will wait until the HANA installation is done and actually the cluster is even created, so i really don't see many options where the DB import and HANA installation/SR setup are being executed in parallel

**Comment by yeoldegrove** [comment]
https://github.com/SUSE/sapnwbootstrap-formula/pull/95 and https://github.com/SUSE/ha-sap-terraform-deployments/pull/811 include a new feature to backup HANA after initial DB import (if SR is enabled).
This should fix this issue.
